### PR TITLE
Analyzer: duplicate certificate in Gateway

### PR DESCRIPTION
**Pull request by zufardhiyaulhaq**
Issue: all common problem should have analyzer
https://github.com/istio/istio/issues/21158

common problem: https://istio.io/docs/ops/common-problems/network-issues/#404-errors-occur-when-multiple-gateways-configured-with-same-tls-certificate

limitation:
- only working with `PILOT_SCOPE_GATEWAY_TO_NAMESPACE` false value. I cannot get this value from anywhere. So cannot decide the logic. since the default value is `false`.

- The current logic is comparing the name since I cannot get the raw certificate and somehow comparing the certificate directly with OpenSSL. So we cannot validate the same certificate that configures in a different mode (example: gateway X via `ServerTLSSettings.credentialName` and gateway Y via `ServerTLSSettings.serverCertificate` and `ServerTLSSettings.privateKey`). 

example:
```
Error [IST0135] (Gateway gateway-01-test-01.istio-system galley/pkg/config/analysis/analyzers/testdata/gateway-duplicate-certificate.yaml:4) Duplicate certificate in multiple gateway, causing 404.
Error [IST0135] (Gateway gateway-01-test-02.istio-system galley/pkg/config/analysis/analyzers/testdata/gateway-duplicate-certificate.yaml:42) Duplicate certificate in multiple gateway, causing 404.
Error [IST0135] (Gateway gateway-01-test-03.default galley/pkg/config/analysis/analyzers/testdata/gateway-duplicate-certificate.yaml:81) Duplicate certificate in multiple gateway, causing 404.
Error [IST0135] (Gateway gateway-02-test-01.istio-system galley/pkg/config/analysis/analyzers/testdata/gateway-duplicate-certificate.yaml:23) Duplicate certificate in multiple gateway, causing 404.
```

[x] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[x] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Comment by zufardhiyaulhaq** [comment]
can we reopen and review this MR?

**Review by esnible — COMMENTED** [review]
The code looks fine.  I can only understand the analysis part.  Before we approve we need two things.

1. Can someone from @istio/wg-networking-maintainers confirm that the inner logic is correct?
2. @zufardhiyaulhaq Can you write instructions to manually introduce and manually correct this config problem, so I can test this validator interactively?

**Comment by zufardhiyaulhaq** [comment]
> The code looks fine. I can only understand the analysis part. Before we approve we need two things.
> 
> 1. Can someone from @istio/wg-networking-maintainers confirm that the inner logic is correct?
> 2. @zufardhiyaulhaq Can you write instructions to manually introduce and manually correct this config problem, so I can test this validator interactively?

It's well documented in Istio docs, https://istio.io/docs/ops/common-problems/network-issues/#404-errors-occur-when-multiple-gateways-configured-with-same-tls-certificate. You have two gateway in the same Istio ingress gateway with the same certificate

File: galley/pkg/config/analysis/msg/messages.yaml
Comment (howardjohn) [inline galley/pkg/config/analysis/msg/messages.yaml:377]: I think this should be a warning. The config is 100% valid, it just doesn't behave as one might expect *in some cases* (due to https://github.com/envoyproxy/envoy/issues/6767).

Also can you link to https://github.com/envoyproxy/envoy/issues/6767 here, so we know to remove this analyzer if/when that issue is resolved?.

I would also change the warning to something like `"Duplicate certificate in multiple gateway, which may cause 404s if clients re-use HTTP2 connections"`. And hopefully the IST0137 doc will have more info about this (like the current istio.io doc)?

File: galley/pkg/config/analysis/msg/messages.yaml
Comment (howardjohn) [inline galley/pkg/config/analysis/msg/messages.yaml:379]: can we put which certificate/gateway caused it?

File: galley/pkg/config/analysis/msg/messages.yaml
Comment (zufardhiyaulhaq) [reply galley/pkg/config/analysis/msg/messages.yaml:377]: sure will do that, refer to the issue as a comment in the code?

Comment (howardjohn) [reply galley/pkg/config/analysis/msg/messages.yaml:377]: Yeah I don't think we want it in the user facing info, just a comment in the code

File: galley/pkg/config/analysis/msg/messages.yaml
Comment (zufardhiyaulhaq) [reply galley/pkg/config/analysis/msg/messages.yaml]: done, putting gateway which caused it

File: galley/pkg/config/analysis/msg/messages.yaml
Comment (zufardhiyaulhaq) [reply galley/pkg/config/analysis/msg/messages.yaml]: this done, will add documentation in the istio.io repository after this merge

File: galley/pkg/config/analysis/msg/messages.yaml
Comment (esnible) [inline galley/pkg/config/analysis/msg/messages.yaml:394]:
```suggestion
    description: "Duplicate certificate in multiple gateways may cause 404s if clients re-use HTTP2 connections."
```

File: galley/pkg/config/analysis/msg/messages.yaml
Comment (esnible) [inline galley/pkg/config/analysis/msg/messages.yaml:395]:
```suggestion
    template: "Duplicate certificate in multiple gateways %v may cause 404s if clients re-use HTTP2 connections."
```

**Review by esnible — APPROVED** [review]

**Review by howardjohn — APPROVED** [review]

**Pull request merged**
